zanata-war/src/main/java/org/zanata/util/HttpUtil.java

```diff
@@ -0,0 +1,106 @@
+/*
+ * Copyright 2015, Red Hat, Inc. and individual contributors as indicated by the
+ * @author tags. See the copyright.txt file in the distribution for a full
+ * listing of individual contributors.
+ *
+ * This is free software; you can redistribute it and/or modify it under the
+ * terms of the GNU Lesser General Public License as published by the Free
+ * Software Foundation; either version 2.1 of the License, or (at your option)
+ * any later version.
+ *
+ * This software is distributed in the hope that it will be useful, but WITHOUT
+ * ANY WARRANTY; without even the implied warranty of MERCHANTABILITY or FITNESS
+ * FOR A PARTICULAR PURPOSE. See the GNU Lesser General Public License for more
+ * details.
+ *
+ * You should have received a copy of the GNU Lesser General Public License
+ * along with this software; if not, write to the Free Software Foundation,
+ * Inc., 51 Franklin St, Fifth Floor, Boston, MA 02110-1301 USA, or see the FSF
+ * site: http://www.fsf.org.
+ */
+package org.zanata.util;
+
+import java.util.Arrays;
+import java.util.LinkedList;
+import java.util.List;
+import javax.servlet.http.HttpServletRequest;
+import javax.ws.rs.HttpMethod;
+
+import org.apache.commons.lang.StringUtils;
+import org.jboss.resteasy.spi.HttpRequest;
+
+import com.google.common.collect.Lists;
+
+/**
+ * Utility class for HTTP related methods.
+ *
+ * @author Patrick Huang
+ *         <a href="mailto:pahuang@redhat.com">pahuang@redhat.com</a>
+ */
+public final class HttpUtil {
+    private final static List<String> HTTP_REQUEST_READ_METHODS = Lists.newArrayList(
+        HttpMethod.GET, HttpMethod.HEAD, HttpMethod.OPTIONS);
+
+    public static final String X_AUTH_TOKEN_HEADER = "X-Auth-Token";
+    public static final String X_AUTH_USER_HEADER = "X-Auth-User";
+
+    public static final String X_FORWARDED_FOR = "X-Forwarded-For";
+    public static final String PROXY_CLIENT_IP = "Proxy-Client-IP";
+    public static final String WL_Proxy_Client_IP = "WL-Proxy-Client-IP";
+    public static final String HTTP_CLIENT_IP = "HTTP_CLIENT_IP";
+    public static final String HTTP_X_FORWARDED_FOR = "HTTP_X_FORWARDED_FOR";
+
+    private static final LinkedList<String> PROXY_HEADERS = Lists
+            .newLinkedList((Arrays.asList(X_FORWARDED_FOR, PROXY_CLIENT_IP,
+                    WL_Proxy_Client_IP, HTTP_CLIENT_IP, HTTP_X_FORWARDED_FOR)));
```

Comment:
If you only use these strings once, I would just put them in here as literals.  Making them public constants increases the size of this class's API.  It's a bit of a random collection of header names, any attribution other than stackoverflow?

http://distinctplace.com/infrastructure/2014/04/23/story-behind-x-forwarded-for-and-x-real-ip-headers/

I think X-Forwarded-For can actually be a list of IP addresses, starting with the client's purported IP address (possibly forged, or inside a LAN with NAT) and hopefully ending with the real IP address which actually connected to our HTTP proxy.  Unless of course Zanata is running without Apache/nginx/haproxy in front of it, in which case we shouldn't be looking at any proxy headers.

Since we're using this for rate limiting, we need to be careful about trusting the headers.  What's to stop someone putting a false header into their requests?  Or perhaps they have a genuine HTTP proxy on their office LAN, the IP addresses of which are irrelevant to us.

Perhaps we should only accept a single optional header name, which should be specified as a system property by the system admin.  `X-Forwarded-For` seems to be pretty standard, so we should make sure we can extract the last IP in the list, as mentioned in the linked page above.